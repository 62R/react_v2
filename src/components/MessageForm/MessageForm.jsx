import style from './MessageForm.module.scss';
import { useEffect, useState } from 'react';
import { Button } from '../Button/Button';
import { Input } from '../Input/Input';
import { Textarea } from '../Textarea/Textarea';

export const MessageForm = ({ submitFn }) => {
  const [author, setAuthor] = useState('');
  const [text, setText] = useState('');

  const [authorIsValid, setAuthorValid] = useState(true);
  const [textIsValid, setTextValid] = useState(true);
  const [buttonState, setButtonState] = useState(true);

  useEffect(() => {
    author && text ? setButtonState(false) : setButtonState(true);
  }, [author, text]);

  const handleSubmit = (evt) => {
    evt.preventDefault();
    if (!buttonState) {
      submitFn(author, text);
      setText('');
      setTextValid(true);
    }
  };

  const changeAuthor = (evt) => {
    setAuthor(evt.target.value);
    setAuthorValid(checkValidation(evt.target.value));
  };

  const changeText = (evt) => {
    setText(evt.target.value);
    setTextValid(checkValidation(evt.target.value));
  };

  const checkValidation = (value) => {
    if (value !== '') {
      return true;
    }
    return false;
  };

  return (
    <form className={style.form} onSubmit={handleSubmit}>
      <Input
        value={author}
        handleChange={changeAuthor}
        isValid={authorIsValid}
        placeholder="Enter your name"
      />
      <Textarea
        value={text}
        handleChange={changeText}
        isValid={textIsValid}
        placeholder="Write some message"
      />
      <Button label="Push" disabled={buttonState} />
    </form>
  );
};
