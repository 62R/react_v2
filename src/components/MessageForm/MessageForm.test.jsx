import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

import { MessageForm } from './MessageForm';
import { act } from 'react-dom/test-utils';

const mockHandler = jest.fn();
let inputs = {};
let button = {};

beforeEach(() => {
  render(<MessageForm submitFn={mockHandler} />);
  inputs = screen.getAllByRole('textbox');
  button = screen.getByRole('button');
});

describe('MessageForm', () => {
  it('render component with snapshot', () => {
    const { asFragment } = render(<MessageForm />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('check validation: Inputs was Empty = false', async () => {
    expect(button).toBeDisabled();
  });

  it('check validation: Name was Empty = false', async () => {
    await act(async () => {
      await userEvent.type(inputs[0], 'Jon{backspace}{backspace}{backspace}');
      await userEvent.type(inputs[1], 'Text');
    });
    expect(button).toBeDisabled();
    expect(inputs[0]).toHaveClass('input_valid_false');
  });

  it('check validation: Text was Empty = false', async () => {
    await act(async () => {
      await userEvent.type(inputs[0], 'Jon');
      await userEvent.type(inputs[1], 'Tex{backspace}{backspace}{backspace}');
    });
    expect(button).toBeDisabled();
    expect(inputs[1]).toHaveClass('textarea_valid_false');
  });

  it('submit form', async () => {
    await act(async () => {
      await userEvent.type(inputs[0], 'Jon');
      await userEvent.type(inputs[1], 'Text');
    });

    await act(async () => {
      await userEvent.click(button);
    });

    expect(mockHandler).toBeCalledTimes(1);
  });

  it('clear Text after submit', async () => {
    await act(async () => {
      await userEvent.type(inputs[0], 'Jon');
      await userEvent.type(inputs[1], 'Text');
    });

    await act(async () => {
      await userEvent.click(button);
    });
    expect(inputs[1]).toHaveValue('');
  });
});
