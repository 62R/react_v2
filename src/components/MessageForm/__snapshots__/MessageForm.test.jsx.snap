// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageForm render component with snapshot 1`] = `
<DocumentFragment>
  <form
    class="form"
  >
    <input
      class="input"
      placeholder="Enter your name"
      type="text"
      value=""
    />
    <textarea
      class="textarea"
      placeholder="Write some message"
    />
    <button
      class="button button_disabled"
      disabled=""
    >
      Push
    </button>
  </form>
</DocumentFragment>
`;
